pkg/util/querylimits/limiter.go

```diff
@@ -0,0 +1,100 @@
+package querylimits
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log/level"
+
+	"github.com/grafana/loki/pkg/util/limiter"
+	logutil "github.com/grafana/loki/pkg/util/log"
+)
+
+type Limiter struct {
+	limiter.CombinedLimits
+}
+
+func NewLimiter(original limiter.CombinedLimits) *Limiter {
+	return &Limiter{
+		CombinedLimits: original,
+	}
+}
+
+// MaxQueryLength returns the limit of the length (in time) of a query.
+func (l *Limiter) MaxQueryLength(ctx context.Context, userID string) (time.Duration, error) {
+	original, err := l.CombinedLimits.MaxQueryLength(ctx, userID)
+	// in theory this error should never happen
+	if err != nil {
+		return 0, err
+	}
+	requestLimits := ExtractQueryLimitsContext(ctx)
+	if requestLimits == nil || requestLimits.MaxQueryLength == 0 {
+		_ = level.Debug(logutil.WithContext(ctx, logutil.Logger)).Log("msg", "using original limit")
+		return original, nil
+	}
+	if time.Duration(requestLimits.MaxQueryLength) > original {
+		_ = level.Error(logutil.WithContext(ctx, logutil.Logger)).Log("msg", "invalid MaxQueryLength", "tenant", userID, "requestLimit", requestLimits.MaxQueryLength, "tenantLimit", original)
+		return 0, fmt.Errorf("invalid MaxQueryLength of %v, tenant limit is %v", requestLimits.MaxQueryLength, original)
+	}
+	_ = level.Debug(logutil.WithContext(ctx, logutil.Logger)).Log("msg", "using request limit", "limit", requestLimits.MaxQueryLength)
+	return time.Duration(requestLimits.MaxQueryLength), nil
+}
+
+func (l *Limiter) QueryTimeout(ctx context.Context, userID string) (time.Duration, error) {
+	original, err := l.CombinedLimits.QueryTimeout(ctx, userID)
+	// in theory this error should never happen
+	if err != nil {
+		return 0, err
+	}
+	requestLimits := ExtractQueryLimitsContext(ctx)
+	if requestLimits == nil || requestLimits.QueryTimeout == 0 {
+		_ = level.Debug(logutil.WithContext(ctx, logutil.Logger)).Log("msg", "using original limit")
+		return original, nil
+	}
+	if time.Duration(requestLimits.QueryTimeout) > original {
+		_ = level.Error(logutil.WithContext(ctx, logutil.Logger)).Log("msg", "invalid QueryTimeout", "tenant", userID, "requestLimit", requestLimits.QueryTimeout, "tenantLimit", original)
+		return 0, fmt.Errorf("invalid QueryTimeout of %v, tenant limit is %v", requestLimits.QueryTimeout, original)
```

Comment:
I don't think this should be here. `QueryTimeout` is called during query evaluation. End-users executing queries should not see this error. We should apply the more restricting limit instead like we've done in the prototype.

I think this should be in another API endpoint that is called by the admin that sets the policies.